### PR TITLE
[pkg/stanza/fileconsumer] Add ability to read files asynchronously

### PR DESCRIPTION
**Pull request by VihasMakwana**
**Description:** Added a new feature gate that enables a thread pool mechanism to respect the `poll_interval` parameter.

Current Scenario:

If a file takes longer than poll_interval to consume, the current implementation would block until it consumes entirely. In other words, it doesn't respect poll_interval.
Improvisation using thread pooling:

In a thread pool model, the backend will queue the files as it proceeds and won't wait for them to consume, all the reading will be asynchronous.

**Link to tracking Issue:** [#18908](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/18908)

**Testing:** Nothing new added, existing ones are modified as per the feature gate

### CODE REVIEW

**Comment by VihasMakwana** [comment]
@djaglowski PR is up for review ;)

File: pkg/stanza/fileconsumer/file_threadpool.go [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+	ctx, cancel := context.WithCancel(context.Background())
+	m.syncLastPollFilesConcurrent(ctx)
+	cancel()
+}
+
+// poll checks all the watched paths for new entries
+func (m *Manager) pollConcurrent(ctx context.Context) {
+	// Increment the generation on all known readers
+	// This is done here because the next generation is about to start
+	m.knownFilesLock.Lock()
```
Comment (VihasMakwana) [inline pkg/stanza/fileconsumer/file_threadpool.go:46]: @djaglowski I _guess_ we can remove locking from this spot? I wasn't sure about it that's why I pushed it.
We update each reader's generation, we don't append/delete the array in any form.

File: pkg/stanza/fileconsumer/file_threadpool.go [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+	"sync"
+
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/fileconsumer/internal/fingerprint"
+)
+
+type readerWrapper struct {
+	reader *reader
+	fp     *fingerprint.Fingerprint
```
Comment (djaglowski) [inline pkg/stanza/fileconsumer/file_threadpool.go:18]: What's the reason for this struct? Is this fingerprint different than the one inside of the reader?

File: pkg/stanza/fileconsumer/file_threadpool.go [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+	m.trieLock.RLock()
+	defer m.trieLock.RUnlock()
+	return m.trie.HasKey(fp.FirstBytes)
+}
+
+func (m *Manager) removePath(fp *fingerprint.Fingerprint) {
+	m.trieLock.Lock()
+	defer m.trieLock.Unlock()
+	m.trie.Delete(fp.FirstBytes)
+}
```
Comment (djaglowski) [inline pkg/stanza/fileconsumer/file_threadpool.go:142]: Should we just move the locking into the trie?

File: pkg/stanza/fileconsumer/util_test.go [hunk 4]
```diff
@@ -84,13 +85,28 @@ func withEmitChan(emitChan chan *emitParams) testManagerOption {
...
+}
+
+func buildTestManager(t testing.TB, cfg *Config, opts ...testManagerOption) (*Manager, chan *emitParams) {
 	tmc := &testManagerConfig{emitChan: make(chan *emitParams, 100)}
 	for _, opt := range opts {
 		opt(tmc)
 	}
 	input, err := cfg.Build(testutil.Logger(t), testEmitFunc(tmc.emitChan))
 	require.NoError(t, err)
+	if tmc.initializeChannel && useThreadPool.IsEnabled() {
```
Comment (djaglowski) [inline pkg/stanza/fileconsumer/util_test.go:101]: We probably don't need to check the feature gate in tests, right? We're in full control via the `withReaderChan` option.

File: pkg/stanza/fileconsumer/file_threadpool.go [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+	reader, err := m.newReaderConcurrent(file, fp)
+	if err != nil {
+		m.Errorw("Failed to create reader", zap.Error(err))
+		return nil, nil
+	}
+	return reader, fp
+}
+
+func (m *Manager) consumeConcurrent(ctx context.Context, paths []string) {
+	m.clearOldReadersConcurrent(ctx)
```
Comment (djaglowski) [inline pkg/stanza/fileconsumer/file_threadpool.go:119]: This appears to be a blocking call, so doesn't it effectively segments all file reading into poll intervals? I think we may still have performance benefits from reading files asynchronously _within_ a poll interval (as opposed to processing all files in a batch before moving to the next batch).

File: pkg/stanza/fileconsumer/file_threadpool.go [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+		}
+	}
+	m.knownFiles = m.knownFiles[i:]
+
+	var lostWG sync.WaitGroup
+	for _, r := range oldReaders {
+		lostWG.Add(1)
+		go func(r *reader) {
+			defer lostWG.Done()
+			r.ReadToEnd(ctx)
```
Comment (djaglowski) [inline pkg/stanza/fileconsumer/file_threadpool.go:167]: I think this is ok because `clearOldReadersConcurrent` is blocking, but if we want to be truly asynchronous, where a new poll may find the same file and begin reading it, then this would potentially result in reading the same content multiple times (once here, and once from a handle found in the next poll).

**Review by djaglowski — COMMENTED** [review]
Thanks @VihasMakwana. This is my initial pass at the review. I'm still trying to wrap my head around some aspects of this implementation but I wanted to leave initial thoughts.

I mentioned [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/17846#issuecomment-1695795805) that I think it's important for us to have benchmarks in our codebase that demonstrate the improvement we want to see from the change, but we also need to have some confidence that we're not meaningfully regressing in other cases such as those where files are typically about the same size.

File: pkg/stanza/fileconsumer/file_threadpool.go (hunk 3 above)
Comment (VihasMakwana) [reply pkg/stanza/fileconsumer/file_threadpool.go:142]: Well, we can. But I didn't go with it, as I was planning to add Trie in the current model (synchronous) as well in the future.

File: pkg/stanza/fileconsumer/file_threadpool.go (hunk 2 above)
Comment (VihasMakwana) [reply pkg/stanza/fileconsumer/file_threadpool.go:18]: Yeah, it's different. Because the one inside the reader can possibly keep growing as we read to the end.
The one inside struct is the one added to trie, when we first read the file.

File: pkg/stanza/fileconsumer/file_threadpool.go (hunk 3 above)
Comment (VihasMakwana) [reply pkg/stanza/fileconsumer/file_threadpool.go:142]: That's a whole different thing, we can discuss it in a different issue.

File: pkg/stanza/fileconsumer/file_threadpool.go (hunk 5 above)
Comment (VihasMakwana) [reply pkg/stanza/fileconsumer/file_threadpool.go:122]: Okay, I have changed it to non-blocking. 
i just submit it to the channels, just like others and update the trie

**Comment by VihasMakwana** [comment]
@djaglowski I have added the benchmark test case.

File: pkg/stanza/fileconsumer/file_threadpool.go [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+		select {
+		case <-ctx.Done():
+			return
+		case wrapper, ok := <-m.readerChan:
+			if !ok {
+				return
+			}
+			r, fp := wrapper.reader, wrapper.trieKey
+			if !m.readToEnd(ctx, r) && !wrapper.close {
+				// Save off any files that were not fully read or if deleteAfterRead is disabled.
```
[the remark this answers is not in the document]
Comment (VihasMakwana) [reply pkg/stanza/fileconsumer/file_threadpool.go:77]: if you see m.readToEnd(), it returns a bool depending on deleteAfterUse and eof.

File: pkg/stanza/fileconsumer/benchmark_test.go [hunk 8]
```diff
@@ -180,3 +189,66 @@ func BenchmarkFileInput(b *testing.B) {
...
+
+	for i := 0; i < 50; i++ {
+		file := openFile(b, filepath.Join(rootDir, fmt.Sprintf("file_%s.log", uuid.NewString())))
+		file.WriteString(uuid.NewString() + strings.Join(logs[:fileSize.sizes[i%2]], "\n") + "\n")
+	}
+}
+
+func BenchmarkFileSizeVarying(b *testing.B) {
+	fileSize := fileSizeBenchmark{
+		sizes: [2]int{b.N * 5000, b.N * 10}, // Half the files will be huge, other half will be smaller
```
[the remark this answers is not in the document]
Comment (VihasMakwana) [reply pkg/stanza/fileconsumer/benchmark_test.go:216]: I think I have misspelled sizes.
those are actually number of logs added to files.
if those two are number of logs, as you said, is that fine by you?

File: pkg/stanza/fileconsumer/file_threadpool.go [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+
+func (m *Manager) updateTrie(fp *fingerprint.Fingerprint, insert bool) {
+	m.trieLock.Lock()
+	defer m.trieLock.Unlock()
+	if insert {
+		m.trie.Put(fp.FirstBytes)
+	} else {
+		m.trie.Delete(fp.FirstBytes)
+	}
+}
```
[the remark this answers is not in the document]
Comment (VihasMakwana) [reply pkg/stanza/fileconsumer/file_threadpool.go:147]: Yeah, that's actually better I guess.

File: pkg/stanza/fileconsumer/benchmark_test.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (VihasMakwana) [reply pkg/stanza/fileconsumer/benchmark_test.go:216]: I probably need to rename it to `logs` or something.

File: pkg/stanza/fileconsumer/file_threadpool.go [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+		} else {
+			break
+		}
+	}
+	m.knownFiles = m.knownFiles[i:]
+
+	for _, r := range oldReaders {
+		if m.isCurrentlyConsuming(r.Fingerprint) {
+			r.Close()
+		} else {
```
[the remark this answers is not in the document]
Comment (VihasMakwana) [reply pkg/stanza/fileconsumer/file_threadpool.go:168]: > What if we are not currently reading the file, but we did read the file again since this reader was created?

I don't think so if that's possible. Because `m.clearOldReadersConcurrent(ctx)` updates the `knownFiles` and clears out old readers ***before*** the readers for current poll cycles are created.

The current `poll()` will begin **only** when the previous poll cycle has submitted all the readers to the thread pool and the trie will be up to date.
So, there's no chance that we'd add any new file to read between the previous `poll()` and current `clearOldReadersConcurrent()`.

File: pkg/stanza/fileconsumer/rotation_test.go [hunk 11]
```diff
@@ -418,6 +418,8 @@ func TestTrackMovedAwayFiles(t *testing.T) {
 	require.NoError(t, err)
 	writeString(t, movedFile, "testlog2\n")
 	operator.poll(context.Background())
+	operator.poll(context.Background())
+	operator.poll(context.Background())
```
[the remark this answers is not in the document]
Comment (VihasMakwana) [reply pkg/stanza/fileconsumer/rotation_test.go:422]: so, you need 3 poll() calls.

File: pkg/stanza/fileconsumer/file_threadpool.go [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+import (
+	"context"
+	"os"
+
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/stanza/fileconsumer/internal/fingerprint"
+)
+
+type readerWrapper struct {
```
[the remark this answers is not in the document]
Comment (VihasMakwana) [reply pkg/stanza/fileconsumer/file_threadpool.go:15]: `readerEnvelope` loops catchy tbh. I have such namings in some client libraries that utilize channels.
Sounds fine to you? We can do either of them tbh

File: pkg/stanza/fileconsumer/file_threadpool.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (djaglowski) [reply pkg/stanza/fileconsumer/file_threadpool.go:168]: > I don't think so if that's possible. Because m.clearOldReadersConcurrent(ctx) updates the knownFiles and clears out old readers before the readers for current poll cycles are created.

But it only removes readers from 3 poll cycles ago, so readers from the previous cycle are still there.

[the remark this answers is not in the document]
Comment (VihasMakwana) [reply pkg/stanza/fileconsumer/file_threadpool.go:168]: Yeah, I agree. `clearOldReaders`  won't touch those readers which are from previous cycles i.e. less than 3 cycles.

[the remark this answers is not in the document]
Comment (VihasMakwana) [reply pkg/stanza/fileconsumer/file_threadpool.go:168]: and we also have a trie check in place, we won't accidentally push the same reader again.

[the remark this answers is not in the document]
Comment (VihasMakwana) [reply pkg/stanza/fileconsumer/file_threadpool.go:168]: one more thing to note,
If a reader is in `knownFiles` array, it would effectively mean that it isn't being consumed at the moment.

File: pkg/stanza/fileconsumer/util_test.go (hunk 4 above)
Comment (VihasMakwana) [reply pkg/stanza/fileconsumer/util_test.go:101]: Actually, we do need this check here or at [line#90](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/25884/files#diff-540e11ad5bede83a0150f6742d29bca27572279b92fb606af37527d5159837ceR90).
Because we don't place a check before calling `withReaderChan` in test cases. That would result in hoge redundant code.

File: pkg/stanza/fileconsumer/benchmark_test.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (VihasMakwana) [reply pkg/stanza/fileconsumer/benchmark_test.go]: @djaglowski I have made number of logs configurable and have added some more params to better scale this tests.

File: pkg/stanza/fileconsumer/file_threadpool.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (djaglowski) [reply pkg/stanza/fileconsumer/file_threadpool.go:173]: > > I don't think so if that's possible. Because m.clearOldReadersConcurrent(ctx) updates the knownFiles and clears out old readers before the readers for current poll cycles are created.
> 
> But it only removes readers from 3 poll cycles ago, so readers from the previous cycle are still there.

I still think this is a problem. When we `ReadToEnd` as the file is discarded, it will start from an out of date offset. This will cause duplicate ingestion of data.

File: pkg/stanza/fileconsumer/util_test.go (hunk 4 above)
Comment (djaglowski) [reply pkg/stanza/fileconsumer/util_test.go:101]: If we have code here that checks the feature gate, then do we need the `withReaderChan` option at all? It seems redundant that we're passing `withReaderChan` to all these tests, but then only using it if the feature gate is enabled. Can we just set up and use the channel based on the feature gate alone?

**Review by djaglowski — COMMENTED** [review]
I don't think this implementation is respecting `max_concurrent_files`. It uses this setting to create the number of workers, which represents the number of files that can simultaneously be _read_. However, we have 3 generations of files sitting open in `knownFiles` without any regard for this setting. The setting is intended to maximize the number of _open_ files, not the number of workers / actively read files.

**Comment by VihasMakwana** [comment]
Okay, I got your point.
How about we keep `max_concurrent_files / 3` as number  of workers? or `max_concurrent_files/2` and keep knownFiles for 2 poll() cycles.
Because,

**Comment by VihasMakwana** [comment]
Okay, I got your point.
How about we keep `max_concurrent_files / 3` as number of workers? or `max_concurrent_files/2` like the current implementation.

File: pkg/stanza/fileconsumer/file_threadpool.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (VihasMakwana) [reply pkg/stanza/fileconsumer/file_threadpool.go:173]: Dan, if you don't mind, can you give me a walk through with an example? I think we'll be able to sort it out fairly quickly then.

File: pkg/stanza/fileconsumer/util_test.go (hunk 4 above)
Comment (VihasMakwana) [reply pkg/stanza/fileconsumer/util_test.go:101]: Now this is a tricky situation.
Let me give a try to another approach I'm currently thinking.

File: pkg/stanza/fileconsumer/file_threadpool.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (djaglowski) [reply pkg/stanza/fileconsumer/file_threadpool.go:173]: Sure, I'll try to be concise so hopefully this is clear enough.

Let's make the situation unrealistically simple and say that one line is written to a file in between each poll cycle. So at each time, we create a reader that reads to the end of the file. Each of these is a reader for the same file, but each is in a different generate and _has a different offset_.

| Time | Actual Content | Offset After Reading |
| --- | --- | --- |
| t0 | `foo0\n` | 5 |
| t1 | `foo0\nfoo1\n` | 10 |
| t2 | `foo0\nfoo1\nfoo2\n` | 15 |
| t3 | `foo0\nfoo1\nfoo2\nfoo3\n` | 20 |

So at `t3`, when we decide to discard the reader from `t0`, if we `ReadToEnd` _on that reader_, we will read `foo1\n`, `foo2\n...` again, even though `foo1\n` was already read at `t1`, `foo2\n` was read at `t2`, etc.

[the remark this answers is not in the document]
Comment (VihasMakwana) [reply pkg/stanza/fileconsumer/file_threadpool.go:173]: I see. What I can do is, I will create a test case which simulates this and check it out. Should be simple.
Thanks 😊

**Comment by VihasMakwana** [comment]
By the way, do you think reducing the lost reader's generation check from 3 cycles to 2 cycles makes sense to you? It might, because if a file is unavailable for 2 straight polls(), it is unlikely it will available in 3rd.

**Comment by djaglowski** [comment]
> By the way, do you think reducing the lost reader's generation check from 3 cycles to 2 cycles makes sense to you? It might, because if a file is unavailable for 2 straight polls(), it is unlikely it will available in 3rd.

Probably, but I recall experiencing increased instability when dropping this in the past. Ultimately, I am hopeful we will find a design that gets away from this notion entirely.

**Comment by VihasMakwana** [comment]
@djaglowski I have re-introduced [batching](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/25884/files#diff-a918db427e06075a5fa059d7893f54556bfa1504726ba77b3491b644a64ab818R59) in this implementation.  Respecting max_concurrent_files skipped my mind, kind apologies.

I have introduced a new helper to take care of copy-truncate duplications [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/25884/files#diff-a918db427e06075a5fa059d7893f54556bfa1504726ba77b3491b644a64ab818R221). 

I have also tweaked the logic to clear old readers to respect the `max_concurrent_files`.

At the beginning of the batch, we close all the `knownFiles` after reading them one last time. 
This will make sure that there are atmost `max_concurrent_files` open files.

**Comment by VihasMakwana** [comment]
In the meantime, I'm working to reduce redundancy in test cases.

File: pkg/stanza/fileconsumer/file_threadpool.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (VihasMakwana) [reply pkg/stanza/fileconsumer/file_threadpool.go:187]: With a tweak in logic and respecting the `max_concurrent_files`, we don't hold readers open for 3 `poll()` cycles anymore. So, I think we're good now.

To discuss the above-mentioned scenario, we remove a reader from `m.knownFiles` if we detect a match [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/stanza/fileconsumer/file.go#L301). 
Also, while copying a reader, we don't copy the `generation`. 
So, every time a reader is ***copied***, we remove the old reader from `m.knownFiles` and create a fresh reader with `generation` set to `0`.

The scenario would look something like this with knownFiles:

Time | Actual Content | Offset After Reading | knownFiles (before poll()) | knownFiles (after poll())
-- | -- | -- | -- | --
t0 | foo0\n | 5 | [] | [reader{offset: 5, generation: 0}]
t1 | foo0\nfoo1\n | 10 | [reader{offset: 5, generation: 0}] | [reader{offset: 10 generation: 0}]
t2 | foo0\nfoo1\nfoo2\n | 15 |  [reader{offset: 15, generation: 0}] | [reader{offset: 15, generation: 0}]
t3 | foo0\nfoo1\nfoo2\nfoo3\n | 20 |  [reader{offset: 20, generation: 0}] | [reader{offset: 20 generation: 0}]

Note: It is assumed that we finish reading asynchronously before the next poll() is executed.

File: pkg/stanza/fileconsumer/file_threadpool.go [hunk 13]
```diff
@@ -0,0 +1,265 @@
...
+}
+
+func (m *Manager) trieDelete(fp *fingerprint.Fingerprint) {
+	m.trieLock.Lock()
+	defer m.trieLock.Unlock()
+	m.trie.Delete(fp.FirstBytes)
+}
+
+func (m *Manager) clearOldReadersConcurrent(ctx context.Context) {
+	m.knownFilesLock.Lock()
```
Comment (VihasMakwana) [inline pkg/stanza/fileconsumer/file_threadpool.go:164]: @djaglowski I have made `clearOldReadersConcurrent()` a kind of blocking call based on the observations made while running benchmarks. 
We wait for all aggregated readers to finish reading before moving forward. 
 
Observations:

- If the number of files matched is significantly higher than `maxConcurrentFiles` then we are likely going to batch the files more than a couple of times.
- This causes `workers()` to overwhelm and they'll all be busy.
- If we try to send already finished readers to the `workers()` anyway, it will cause a bottleneck, and benchmarks show a significant delay in receiving logs.
- It's better to create separate goroutines to process finished readers and let the `workers()` loosen themselves up for the upcoming `poll()` cycle.
- In this way, we respect `max_concurrent_files`, `workers()` are happy, all's good. The benchmark performs as expected.

**Comment by VihasMakwana** [comment]
@djaglowski does the overall PR look stable to you, as it respects current parameters and passes all the test cases?
 If yes, I will shift my focus to fixing the pipeline now. Open to any thoughts/opinions as always.

**Comment by VihasMakwana** [comment]
@djaglowski ping ^

**Review by djaglowski — COMMENTED** [review]
@VihasMakwana, thanks for iterating on this. It seems like the design has changed quite a bit and I'm struggling to wrap my head around all the nuances. As much as I'd like to see this move forward, I'm uncomfortable with how fluid the design seems to be and I lack confidence that we're making the right changes. I think a large part of the problem is that we're having to rework several overlapping concerns at the same time. Do you see any opportunities to separate concerns into individual PRs?

**Comment by VihasMakwana** [comment]
Okay. On a quick note, I do think making individual PRs will help moving forward.

I can think of following:
1. a PR for file threadpool ability. (a minimalist design, we will expand it later on once we are confident enough). I think we should it as an internal package?
2. a PR for complete functionality, after 1st PR is merged and we're confident to move forward. 
3. Test cases, enabling the feature gate, integrating with existing feature.

**Comment by djaglowski** [comment]
> a PR for file threadpool ability. (a minimalist design, we will expand it later on once we are confident enough). I think we should have it as an internal package?

I'm open to this, though I'm not sure I understand exactly what you have in mind. 

--- 

More generally, I think what we need is not necessary a direct path to the design used in this PR, but rather a path that changes the way we manage other concerns in a way that this PR does not need to worry about them.

For example, we have all these different sets of files, most of which don't necessarily need to be a particular concern of an asynchronous solution.
1. Files we have found which need to be opened, fingerprinted and cross-referenced with files we've previously seen.
2. Files we are currently reading.
3. Files we recently read and have kept open in case they become "lost".
4. Files we are reading again, because they became "lost".
5. Files we have already read and closed, but have kept around so we know where to offset to if we see them again.

In other words, can we redesign the way we manage any of these in such a way that it is agnostic to the implementation of the others?

**Comment by VihasMakwana** [comment]
Okay. If I understand this correctly, what we require is a path of restructuring the PR which addresses individual concerns independently, without relying on others as you said.
If that's true, then I can work on it. Should be possible.
